core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.java

```diff
@@ -853,15 +865,24 @@ protected void getRDF(HttpUriRequest method, RDFHandler handler, boolean require
 				RDFFormat format = RDFFormat.matchMIMEType(mimeType, rdfFormats)
 						.orElseThrow(() -> new RepositoryException(
 								"Server responded with an unsupported file format: " + mimeType));
-				// Check if we can pass through to the output stream directly
+				// Check if we can pass through to the writer directly
 				if (handler instanceof RDFWriter) {
 					RDFWriter rdfWriter = (RDFWriter) handler;
 					if (rdfWriter.getRDFFormat().equals(format)) {
-						OutputStream out = rdfWriter.getOutputStream().orElse(null);
-						if (out != null) {
-							InputStream in = response.getEntity().getContent();
-							IOUtils.copy(in, out);
-							return;
+						if (rdfWriter instanceof CharSink) {
```

Comment:
I think the following block is quite long and a bit hard to read. Would it be possible to split it up with some private methods? Or replacing the instanceof conditionals with use of the Strategy/Visitor Pattern? I tend to prefer it to get better compile-time checks.

Reply:
I didn't quite use a strategy pattern in the end but I've managed to simplify things a little, have another look.

Reply:
Great job! Much better now!